### PR TITLE
[7.2] Add polyfill for spl_object_id()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
As [discussed on php-internals](https://externals.io/message/99751) and implemented in https://github.com/php/php-src/pull/2611.
Implementation [borrowed from VarDumper](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/VarDumper/Cloner/VarCloner.php#L300).

### CODE REVIEW

File: src/Php72/Php72.php [hunk 1]
```diff
@@ -80,4 +83,31 @@ public static function php_os_family()
 
         return isset($map[PHP_OS]) ? $map[PHP_OS] : 'Unknown';
     }
+
+    public static function spl_object_id($object)
+    {
+        if (null === self::$hashMask) {
```
Comment (lyrixx) [inline src/Php72/Php72.php]: You can use an early return here.

Comment (nicolas-grekas) [inline src/Php72/Php72.php]: not sure it would help, L108 and L111 are not the same, so that would require duplicating L111, is that what you mean?

Comment (keradus) [inline src/Php72/Php72.php]: actually hashMask is calculated without usage of input param here`$object`. maybe extract it to separated function and then 

```php
public static function spl_object_id($object)
{
    return self::getHashMask() ^ hexdec(substr(spl_object_hash($object), self::$hashOffset, PHP_INT_SIZE));
}
```
?

Comment (nicolas-grekas) [inline src/Php72/Php72.php]: I prefer raw perf, thus prevent an extra call here

Comment (nicolas-grekas) [inline src/Php72/Php72.php]: here we are, init logic split in separate function, called only once

File: tests/Php72/Php72Test.php [hunk 2]
```diff
@@ -48,4 +48,18 @@ public function testPhpOsFamily()
...
           $this->assertSame(PHP_OS_FAMILY, p::php_os_family());
     }
+
+    /**
+     * @covers Symfony\Polyfill\Php72\Php72::spl_object_id
+     */
+    public function testSplObjectId()
+    {
+        $obj = new \stdClass();
+        $id = spl_object_id($obj);
```
Comment (keradus) [inline tests/Php72/Php72Test.php]: please test that calling `spl_object_id` for non-object works the same as native function

**Comment by stof** [comment]
@nicolas-grekas I suppose that VarDumper can then be migrated to use the new `spl_object_id` function, right ?

**Comment by stof** [comment]
note: merging this PR must wait until the PR is merged in PHP itself. We don't want to merge the polyfill if they reject the PR.

File: src/Php72/Php72.php [hunk 3]
```diff
@@ -80,4 +83,36 @@ public static function php_os_family()
...
+            self::initHashMask();
+        }
+
+        return self::$hashMask ^ hexdec(substr(spl_object_hash($object), self::$hashOffset, PHP_INT_SIZE));
+    }
+
+    private static function initHashMask()
+    {
+        $obj = (object) array();
+        self::$hashOffset = 16 - PHP_INT_SIZE;
```
Comment (stof) [inline src/Php72/Php72.php]: if you want to improve perf even further, you could even fully qualify these constants (as done for `\PHP_VERSION_ID`). Not sure it is worth it though (this code is called only once)

File: src/Php72/Php72.php (hunk 1 above)
Comment (lyrixx) [inline src/Php72/Php72.php]: :+1:

File: tests/Php72/Php72Test.php (hunk 2 above)
Comment (nicolas-grekas) [inline tests/Php72/Php72Test.php]: done

File: src/Php72/Php72.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Php72/Php72.php]: done by removing $hashOffset

**Comment by nicolas-grekas** [comment]
@stof VarDumper can use this polyfill only in master, because of HHVM compat...

**Comment by robfrawley** [comment]
The `spl_object_id()` function now merged upstream in https://github.com/php/php-src/commit/5097e2ee13de12b4445b4123e1554c0733c6853c (also see https://github.com/php/php-src/pull/2611#issuecomment-319717081) so this can now be safely merged here, as well.

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by robfrawley** [comment]
@fabpot Sorry to comment on a closed PR, but I don't think this is worth a new issue: it would be great to get a release tagged that includes this PR (I hate having to include `dev-master` in my composer files). :-)

**Comment by fabpot** [comment]
@robfrawley Done, 1.5 is out now.

**Pull request merged**